ships/0031-shipwright-trigger.md

```diff
@@ -0,0 +1,383 @@
+| Tekton Pipeline Controller  | Pipeline          | Watching Tekton objects and reacting   |
+
+All the modifications needed are on the `Build` resource under `.spec.trigger` attribute, making this new feature backwards compatible with `v1alpha1`.
+
+## WebHook Events
+
+The following image represents the actors involved in the WebHook request workflow, and shows the order on which the actors are actionated:
+
+![Shipwright Trigger WebHook Schematics](assets/0031-shipwright-trigger-webhook.drawio.png)
+
+The start of the workflow (1) is at the Shipwright Trigger instance, it will be watching over `Build` objects using a regular Kubernetes Controller. Since the Trigger Controller is aware of all `Build` instances, it creates an inventory of `Builds` carrying trigger rules (`.spec.trigger` attribute), and therefore, it's able to search for `Build` objects. The Trigger will only start listening for WebHook requests when the controller has synchronized, so it has the inventory full beforehand.
+
+The WebHook request (2) carries on information about the repository, so Shipwright Trigger can filter out the `Build` objects representing that specific repository, and by inspecting the event details determine if it should create a `BuildRun` object. 
+
+For instance, consider the following snippet of a GitHub Push event:
+
+
+```json
+{
+    "ref": "refs/heads/main",
+    "before": "before-commit-id",
+    "repository": {
+   	 "full_name": "organization/repository",
+   	 "html_url": "https://github.com/organization/repository"
+    },
+    "head_commit": {
+   	 "message": "commit message",
+   	 "author": {
+   		 "name": "Author's Name"
+   	 },
+   	 "id": "commit-id",
+   	 "timestamp": "2022-02-18T11:02:02.635313+01:00"
+    }
+}
+```
+
+Note the snippet attribute `.repository.htm_url` can be compared to Build's `.spec.source.url` in order to determine which `Build` objects are affected by the event. Likewise, we can employ `.ref` to determine which branch is affected, and later asseting if it's part of the Build `.spec.trigger.when[].branches` entries.
+
+Depending on the event type, it will need to inspect `.head_commit.id` as well, so in other words, the events offer more than enough information to fulfill the needs of triggering Shipwright Builds (3).
+
+### Git Service Providers Support
+
+Shipwright Trigger will initially ship with [GitHub support][githubGoGithub], users will be able to configure the WebHook endpoint with additional [secret-token verification][githubWebHookSecretToken], in order to trigger Shipwright Builds accordingly.
+
+The GitHub support is based on a common interface, on which we will support more service providers in the near future, like for instance GitLab and others. Therefore, the interface semantics must be able to implement the following use cases:
+
+- Parse the WebHook request payload identifying the specific type of event
```

Comment:
An alternative would be that we serve different paths for each vendor that we support, `/event/github` vs `/event/gitlab`. For usability reasons, it might make sense if we also extend the status of the Build. It will probably be impossible to (always) determine the externally exposed URL of the service especially if it's not just the Ingress but cluster-external components route requests. So, we will need a configuration option where we are told about it. But then we can use that and put the URL of the endpoint into the Build status from where the user who created the Build can copy it into the GitHub / GitLab etc UI.

Reply:
we handle it in the same fashion as @SaschaSchwarze0 describes with openshift builds and that has proven maintainable / successful 

+1

Reply:
Let me clarify this point a bit better, @SaschaSchwarze0. The need to parse the GitHub event is to determine which exact event is in the request, as you can see on the [GitHub WebHook documentation][ghWebHookEventsAndPayloads] there are a number of those.

So, in one hand, we always need to parse GitHub WebHook request payloads, but in the other we can easily extend the triggers support to more intersting [use-cases][ghaWorkflowRun], like for instance [GitHub Actions Workflow][otaviofCommentGHA].

> An alternative would be that we serve different paths for each vendor that we support, `/event/github` vs `/event/gitlab`.

That's the current proposal, Sascha. Indeed, we can extend the WebHook endpoints by support different request patterns differentiating providers like GitHub, GitLab, etc.

> For usability reasons, it might make sense if we also extend the status of the Build.

Yes, most defintely 👍 And it's also part of the current proposal, suggested by @gabemontero [early on][gabeCommentStatus].

> It will probably be impossible to (always) determine the externally exposed URL of the service especially if it's not just the Ingress but cluster-external components route requests.

The external endpoint URL is responsibility of the cluster administrator, this persona will know in advance the desired URL and create the necessary configuration to expose Shipwright Trigger.

I touched this subject as part of the [*Risks and Mitigations* section][risksAndMitigations] which describes the different ways of bringing external traffic into the cluster. For instance, the Ingress Controller as you mentioned, also options like service mesh and OpenShift routes.

So, while the Shipwright Trigger won't be able to know the exact URL, in order to determine the exact circumstance the WebHook request was issued, GitHub offers a [very good amount of information][ghWebHookEventsAndPayloads] to determine repository, originating user, and a lot more.

In other words we need to extend to work invested on parsing the request payload to extract more meta information needed.

> So, we will need a configuration option where we are told about it. But then we can use that and put the URL of the endpoint into the Build status from where the user who created the Build can copy it into the GitHub / GitLab etc UI.

Given the security implications explained before, the persona responsible to expose the cluster for external requests will provide the final URL. On this [video][youtubeWebHookDemo] you can see how the configuration would happen.


[risksAndMitigations]: https://github.com/shipwright-io/community/blob/033fc66198c2f857cce92cc040ef8d32b89958b2/ships/0031-shipwright-trigger.md#risks-and-mitigations
[ghaWorkflowRun]: https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#workflow_run
[ghWebHookEventsAndPayloads]: https://docs.github.com/en/developers/webhooks-and-events/webhooks/webhook-events-and-payloads
[youtubeWebHookDemo]: https://youtu.be/c3CZlOXEhMY
[gabeCommentStatus]: https://github.com/shipwright-io/community/pull/68#discussion_r839832981
[otaviofCommentGHA]: https://github.com/shipwright-io/community/pull/68#discussion_r836401136